core/src/state/world_state.rs

```diff
@@ -65,44 +68,55 @@ impl StateHandle {
             .write()
             .expect("Could not acquire world_state lock.")
     }
-
-    /** Block asynchronously until the given sequence number is reached. */
-    pub async fn wait_for_seq(&self, sequence: u64) {
-        if self.state().logical_time >= sequence {
-            return;
-        }
-        let receiver = { self.write_state().get_listener(sequence) };
-        receiver.notified().await;
-    }
 }
 
 #[derive(Default, Debug)]
 pub struct WorldState {
     logical_time: u64,
     pub clusters: BTreeMap<ClusterName, ClusterState>,
-    listeners: HashMap<u64, Arc<Notify>>,
+    listeners: RwLock<BTreeMap<u64, Sender<()>>>,
```

Comment:
I think we discussed this offline early in the design, but given where things landed, does it still make sense to have double-nested `RwLock`s? The way I see it, the "owned" value is the `StateHandle`, which already has `WorldState` in an `RwLock`. If we expose `StateHandle::write_state`, we could leave this as a `HashMap` and simplify the code a bit.

Reply:
I agree, but I don't like exposing write_state like that tbh, I've put wait_for_seq on StateHandle, which I think is better.